dashboard/app/controllers/report_abuse_controller.rb

```diff
@@ -109,4 +84,116 @@ def report_abuse_form
     }
   end
+
+  # API routes, moved from legacy channels_api.rb.
+
+  # GET /v3/channels/:channel_id/abuse
+  # Get an abuse score.
+  def show_abuse
+    return head :unauthorized unless project_validator?
+
+    begin
+      value = Projects.get_abuse(params[:channel_id])
+    rescue ArgumentError, OpenSSL::Cipher::CipherError
+      raise ActionController::BadRequest.new, "Bad channel_id"
+    end
+    render json: {abuse_score: value}
+  end
+
+  # DELETE /v3/channels/:channel_id/abuse
+  # POST /v3/channels/:channel_id/abuse/delete
+  # Clear an abuse score. Requires project_validator permission
+  def destroy_abuse
+    return head :unauthorized unless project_validator?
+
+    channel_id = params[:channel_id]
+
+    begin
+      value = Projects.new(get_storage_id).reset_abuse(channel_id)
+    rescue ArgumentError, OpenSSL::Cipher::CipherError
+      raise ActionController::BadRequest.new, "Bad channel_id"
+    end
+    render json: {abuse_score: value}
+  end
+
+  # Non-actions, public methods so they can be tested easier.
+  # The methods below are in this controller because they depend on the
+  # storage_id helper, which has dependencies on current_user.
+
+  # Increment an abuse score.
+  # Was POST /v3/channels/:channel_id/abuse
+  def update_channel_abuse_score(channel_id)
+    # Reports of abuse from verified teachers are more reliable than reports
+    # from students so we increase the abuse score enough to block the project
+    # with only one report from a verified teacher.
+    #
+    # Temporarily ignore anonymous reports and only allow verified teachers
+    # and signed in users to report.
+    restrict_reporting_to_verified_users = DCDO.get('restrict-abuse-reporting-to-verified', true)
+    amount =
+      if current_user&.verified_teacher?
+        20
+      elsif current_user && !restrict_reporting_to_verified_users
+        10
+      else
+        0
+      end
+    begin
+      value = Projects.new(get_storage_id).increment_abuse(channel_id, amount)
+    rescue ArgumentError, OpenSSL::Cipher::CipherError
+      raise ActionController::BadRequest.new, "Bad channel_id"
+    end
+    value
+  end
+
+  # Update all assets for the given channelId to have the provided abuse score
+  # Was PATCH /v3/(animations|assets|sources|files|libraries)/:channel_id?abuse_score=<abuse_score>
+  def update_file_abuse_score(endpoint, encrypted_channel_id, abuse_score)
+    return if abuse_score.nil?
+
+    buckets = get_bucket_impl(endpoint).new
+
+    begin
+      files = buckets.list(encrypted_channel_id)
+    rescue ArgumentError, OpenSSL::Cipher::CipherError
+      raise ActionController::BadRequest.new, "Bad channel_id"
+    end
+    files.each do |file|
+      break unless can_update_abuse_score?(endpoint, encrypted_channel_id, file[:filename], abuse_score)
+      buckets.replace_abuse_score(encrypted_channel_id, file[:filename], abuse_score)
+    end
+
+    abuse_score
+  end
+
+  private
+
+  def project_validator?
```

Comment:
you shouldn't need this any more if you move the auth checks to CanCan, but if you do still need it then you can use `user.project_validator?` sneakily defined here: https://github.com/code-dot-org/code-dot-org/blob/37a813fab6a96821b38260f032f50dbae87ebe69/dashboard/app/models/concerns/user_permission_grantee.rb#L49-L51

Reply:
ah, nice. I think I do still need this to support the logic in `update_file_abuse_score`, but I've updated it to use `current_user.project_validator?`